### ISSUE TITLE
Analyzer proposal: Discourage `== null` and `!= null` syntax

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#56859]
The `== null` and `!= null` syntax in C# is vulnerable to allowing dead checks against struct types and even misleading callers when the types being tested have overridden these operators.

C# has `is null`, `is not null` and `is object` syntax that is immune to all these deficiencies, but most C# developers are not familiar with the preferred syntax and most code is written using the discouraged syntax.

Analyzers could call these out and offer bulk code fixes to switch to the newer syntax. 

In fact I've already done it here: https://github.com/AArnott/CSharpIsNull

Is there interest in taking this into the SDK?

**Comment by CyrusNajmabadi** [comment]
> C# has is null, is not null and is object syntax that is immune to all these deficiencies, but most C# developers are not familiar with the preferred syntax and most code is written using the discouraged syntax.

I don't believe this is discouraged syntax.  Indeed, for some types it is the way to properly do a test.  For example, ImmutableArray overrides these operators so you can do:

```c#
void Foo(ImmutableArray<string> array)
{
    if (array == null) ...
```

in particular 'is null' won't work here, and 'is not object' will give the wrong answer.

**Comment by AArnott** [comment]
`array == null` _works_ with `ImmutableArray<T>`? Wow. I wouldn't have guessed it (even though I wrote it originally). Comparing a struct to null is usually a bug. I would have used `IsDefault` on it or whatever that property is called.

**Comment by CyrusNajmabadi** [comment]
> even though I wrote it originally

:D

> Comparing a struct to null is usually a bug. 

I think teh intent here was to make ImmutableArray a near drop-in replacement for an actual array.  With an actual array you can have null checks, so we wanted to maintain that to prevent churn when using this.  Yes, it's a struct.  But it's a struct to get a no-overhead wrapper over a reference type.  So in that regard it really is intending to feel as close to a ref-type as possible.

**Comment by myokeeh** [comment]
> * `T? a is T` -> `a is not null`

> * Sam: what about `is object` or `is {}`? Should this be offered for these cases?
>   
>   * Everyone: yes please
>   * Cyrus: note that we should not offer this for `is {} identifier` as that is not the same thing

I just switched over to C#10 on a project and noticed a bunch of `IDE0150 Null check can be clarified` which points to a missing page: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0150

Why is `is not null` preferred over `is SomeT`? Is it more efficient?

**Comment by CyrusNajmabadi** [comment]
> Why is is not null preferred over is SomeT

No. It's preferred as it's clearer what the purpose of the check is. For example:

```c#
string s = ...
if (s is string) ...
```

This code is less clear. Given that s is a string, testing to see if it's a string looks superfluous. In this case, the only thing that is actually being tested is not the type, but just that the variable is non-null. 

Since that's all that is being validated, it's clearer and more idiomatic to just actually simplify the check to just be: `if (s is not null)`. Now the code actually reflects exactly what is being tested and is entirely clear about that.

**Comment by elachlan** [comment]
This rule would be super useful. Winforms has a long running issue to manually fix occurrences and this analyzer would allow resolving the issue in minutes instead of multiple PRs and manual/regex changes.

**Comment by AArnott** [comment]
@elachlan Can you use the analyzer and code fix that I wrote? (Linked in the description)?

**Comment by Youssef1313** [comment]
> iff the type of a does not overload the equality operator. The only exception to the does not overload the equality operator rule is the string type as that is sealed and we know what its implementation does.

@jmarolf There are other types in the BCL, e.g, `System.Xml.Linq.XName`. Should we consider anything coming from corlib instead of `System.String` only?

**Comment by Rasmus715** [comment]
I'm rooting for this one as well!

**Comment by santo998** [comment]
Any update on this?
